### PR TITLE
BB-337 Implement metrics for Cold Storage Archive and Restore

### PR DESCRIPTION
**Pull request by nicolas2bert**

### CODE REVIEW

**Comment by nicolas2bert** [comment]
ping

**Comment by nicolas2bert** [comment]
@bert-e reset

File: extensions/utils/metricsUtil.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+function getStatusLabel(err) {
```
Comment (francoisferrand) [inline extensions/utils/metricsUtil.js]: do we really need this?
"aggregating" by status code 'range' can be done at presentation time (in prometheus queries), and there are not that many code anyway ; yet it may useful at times to be able to dig and find the metric for a specific code ?

File: monitoring/lifecycle/dashboard.py [hunk 2]
```diff
@@ -271,7 +271,7 @@ def kafka_row(topic, op):
             layout.row([lifecycle_global_s3_requests], height=10),
             layout.row(lifecycle_global_s3_error_rates, height=4),
             RowPanel(title="Kafka"),
-            layout.row(kafka_row("Expiration Bucket Task", "BucketTopic"), height=10),
+            layout.row(kafka_row("Lifecycle Bucket Task", "BucketTopic"), height=10),
             layout.row(kafka_row("Expiration Object Task", "ObjectTopic"), height=10),
```
Comment (francoisferrand) [inline monitoring/lifecycle/dashboard.py:275]: That one should be renamed as well?

File: extensions/lifecycle/LifecycleMetrics.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 const { ZenkoMetrics } = require('arsenal').metrics;
+const { getStatusLabel } = require('../utils/metricsUtil');
 
 const LIFECYCLE_LABEL_ORIGIN =  'origin';
 const LIFECYCLE_LABEL_OP = 'op';
```
Comment (francoisferrand) [inline extensions/lifecycle/LifecycleMetrics.js]: It would be useful to have "duration" metrics as well: e.g.
* the duration of individual tasks (processing of an object by GC...)
* duration of bucket scan
* latency/delay between an "event" (expiration....) and the time when we actually apply the transition

File: extensions/gc/GarbageCollectorMetrics.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+const gcS3Operations = ZenkoMetrics.createCounter({
+    name: 'gc_s3_operations',
+    help: 'Total number of S3 operations by the garbage collector processes',
+    labelNames: [
+        GC_LABEL_ORIGIN,
+        GC_LABEL_OP,
+        GC_LABEL_STATUS,
+    ],
+});
```
Comment (francoisferrand) [inline extensions/gc/GarbageCollectorMetrics.js:15]: GC may be called for reason other than cold storage expiration ; so we may want to also have some specific metric for restore expiration.

File: monitoring/lifecycle/dashboard.py [hunk 5]
```diff
@@ -271,7 +271,7 @@ def kafka_row(topic, op):
             layout.row([lifecycle_global_s3_requests], height=10),
             layout.row(lifecycle_global_s3_error_rates, height=4),
```
Comment (francoisferrand) [inline monitoring/lifecycle/dashboard.py:272]: dashboard will be updated in a separate PR ?

File: monitoring/lifecycle/dashboard.py (hunk 2 above)
Comment (nicolas2bert) [reply monitoring/lifecycle/dashboard.py:275]: No, the metrics collected are specific to expiration.

File: monitoring/lifecycle/dashboard.py (hunk 5 above)
Comment (nicolas2bert) [reply monitoring/lifecycle/dashboard.py:272]: yes

File: extensions/utils/metricsUtil.js (hunk 1 above)
Comment (nicolas2bert) [reply extensions/utils/metricsUtil.js]: Indeed. I am ok removing this aggregation

File: extensions/gc/GarbageCollectorMetrics.js (hunk 4 above)
Comment (nicolas2bert) [reply extensions/gc/GarbageCollectorMetrics.js:15]: Is leveraging the origin label not ideal? (e.g. `GarbageCollectorMetrics.onS3Request(log, 'getMetadata', 'archive', err);` )

File: extensions/utils/metricsUtil.js (hunk 1 above)
Comment (nicolas2bert) [reply extensions/utils/metricsUtil.js]: Since lifecycle dashboards are already using the "aggregating by status code rage" logic, I see this more as a tech depth that will be tackled by https://scality.atlassian.net/browse/BB-342.

File: extensions/lifecycle/LifecycleMetrics.js (hunk 3 above)
Comment (nicolas2bert) [reply extensions/lifecycle/LifecycleMetrics.js]: This deserves a ticket for itself: https://scality.atlassian.net/browse/BB-343

File: extensions/utils/metricsUtil.js (hunk 1 above)
Comment (francoisferrand) [reply extensions/utils/metricsUtil.js]: TechDebt indeed, though very quick to fix : i.e. use the 'raw' status instead of calling `getStatusLabel` in the code, and replace each of these filters in dashboard (e.g. `="2xx"` with `=~"2.."` for each of the ranges 2/3/4/5)

File: extensions/gc/GarbageCollectorMetrics.js (hunk 4 above)
Comment (francoisferrand) [reply extensions/gc/GarbageCollectorMetrics.js:15]: indeed, i missed that (maybe after seeing the origin being hardcoded in other use cases...)

File: monitoring/lifecycle/dashboard.py (hunk 5 above)
Comment (francoisferrand) [reply monitoring/lifecycle/dashboard.py:272]: 😕
this breaks the kind of "atomicity" we get from having the dashboards in the same repo as the code (and thus the safety of having both the dashboard and matching metrics in the product)

File: extensions/lifecycle/LifecycleMetrics.js (hunk 3 above)
Comment (francoisferrand) [reply extensions/lifecycle/LifecycleMetrics.js]: ok

File: monitoring/lifecycle/dashboard.json [hunk 6]
```diff
@@ -2045,7 +2045,7 @@
       "targets": [
         {
           "datasource": null,
-          "expr": "sum(rate(lifecycle_s3_operations{status=\"2xx\",op=\"deleteObject\",namespace=\"${namespace}\",job=\"${job_lifecycle_object_processor}\"}[$__rate_interval]))",
+          "expr": "sum(rate(lifecycle_s3_operations{status=\"200\",op=\"deleteObject\",namespace=\"${namespace}\",job=\"${job_lifecycle_object_processor}\"}[$__rate_interval]))",
```
Comment (francoisferrand) [inline monitoring/lifecycle/dashboard.json:2048]: same here: maybe we should have `=~"2.."` ?

File: monitoring/lifecycle/dashboard.py [hunk 7]
```diff
@@ -66,9 +66,9 @@ def s3_request_error_rate_expr(process, job, code):
     divsLabel = 'namespace="${namespace}"'
 
     if code is not None:
-        divdLabel += f',status="{code}"'
+        divdLabel += f',status=~"{code}"'
     else:
-        divdLabel += ',status!="2xx"'
+        divdLabel += ',status!="200"'
```
Comment (francoisferrand) [inline monitoring/lifecycle/dashboard.py:71]: do we only ever use 200 on success?
otherwise we should have !~"2.." in such case...

File: monitoring/lifecycle/dashboard.py [hunk 8]
```diff
@@ -110,8 +110,8 @@ def s3_request_error_rates(process=None, job=None):
 
 
 def s3_deletion_request_time_series(op):
-    successLabel = f'status="2xx",op="{op}",namespace="{"${namespace}"}",job="{"${job_lifecycle_object_processor}"}"'
-    errorLabel = f'status!="2xx",op="{op}",namespace="{"${namespace}"}",job="{"${job_lifecycle_object_processor}"}"'
+    successLabel = f'status="200",op="{op}",namespace="{"${namespace}"}",job="{"${job_lifecycle_object_processor}"}"'
+    errorLabel = f'status!="200",op="{op}",namespace="{"${namespace}"}",job="{"${job_lifecycle_object_processor}"}"'
```
Comment (francoisferrand) [inline monitoring/lifecycle/dashboard.py:114]: same here?

**Review by francoisferrand — APPROVED** [review]

File: monitoring/lifecycle/dashboard.py (hunk 7 above)
Comment (nicolas2bert) [reply monitoring/lifecycle/dashboard.py:71]: Yes, it is either an error status code (3xx, 4xx, 5xx) or 200.

**Review by welansari — APPROVED** [review]

**Comment by nicolas2bert** [comment]
@bert-e approve
